src/cmd/cli/command/commands.go

```diff
@@ -859,58 +831,92 @@ var composeUpCmd = &cobra.Command{
+					term.Warnf("failed to wait for service status: %v", err) // TODO: don't print in Go-routine
 				}
+			} else {
+				cancelTail(errCompleted)
 			}
 		}()
 
-		if err := waitServiceState(ctx, defangv1.ServiceState_SERVICE_COMPLETED, serviceInfos); err != nil && !errors.Is(err, context.Canceled) {
-			if errors.Is(err, ErrDeploymentFailed) {
-				term.Warn("Deployment FAILED. Service(s) not running.")
+		// show users the current streaming logs
+		tailSource := "all services"
+		if deploy.Etag != "" {
+			tailSource = "deployment ID " + deploy.Etag
+		}
+
+		term.Info("Tailing logs for", tailSource, "; press Ctrl+C to detach:")
+		tailParams := cli.TailOptions{
+			Etag:  deploy.Etag,
+			Since: since,
+			Raw:   false,
+		}
+
+		// blocking call to tail
+		if err := cli.Tail(ctx, client, tailParams); err != nil {
+			term.Debugf("Tail failed with %v", err)
+			if connect.CodeOf(err) == connect.CodePermissionDenied {
+				// If tail fails because of missing permission, we wait for the deployment to finish
+				term.Warn("Unable to tail logs. Waiting for the deployment to finish.")
+				<-ctx.Done()
+			} else if !errors.Is(ctx.Err(), context.Canceled) {
+				return err // any error other than cancelation
+			}
+
+			// Tail got canceled; if it was by anything other than completion, prompt to show debugger
+			if !errors.Is(context.Cause(ctx), errCompleted) {
+				var failedServices []string
+				var errDeploymentFailed cli.ErrDeploymentFailed
+				if errors.As(context.Cause(ctx), &errDeploymentFailed) {
+					term.Warn(errDeploymentFailed)
+					failedServices = []string{errDeploymentFailed.Service}
+				} else {
+					term.Warn("Deployment is not finished. Service(s) might not be running.")
+					// TODO: some services might be OK and we should only debug the ones that are not
+				}
 
 				if !nonInteractive {
+					// TODO: skip this for BYOC
 					var aiDebug bool
 					if err := survey.AskOne(&survey.Confirm{
 						Message: "Would you like to debug the deployment with AI?",
 						Help:    "This will send logs and artifacts to our backend and attempt to diagnose the issue and provide a solution.",
 					}, &aiDebug); err != nil {
 						term.Debugf("failed to ask for AI debug: %v", err)
 					} else if aiDebug {
-						if err := cli.Debug(ctx, client, deploy.Etag, project.WorkingDir); err != nil {
+						// TODO: use the WorkingDir of the failed service, might not be the project's root
+						if err := cli.Debug(ctx, client, deploy.Etag, project.WorkingDir, failedServices); err != nil {
 							term.Debugf("failed to debug deployment: %v", err)
 						}
 					}
 				}
-
-				return err // return the error from waitServiceState
-			} else {
-				term.Warnf("failed to wait for service status: %v", err)
+				return err
 			}
-
-			wg.Wait() // Wait for tail ctrl + c
 		}
-		cancelTail()
-		wg.Wait() // Wait for tail to finish
 
-		printEndpoints(serviceInfos)
+		printEndpoints(deploy.Services)
```

Comment:
The service states aren't updated in deploy.Services so won't be showing the proper states.

Reply:
ah, so yeah that kind of mutation is not great™ so if we want to show those states we'll need a different solution

Reply:
I saw the new fix.